### ISSUE TITLE
Verify the order of the sampling functions

### ISSUE THREAD

**Issue by seananderson** [ss3sim/ss3sim#181]

**Comment by kellijohnson** [comment]
@colemonnahan I am going through some models / testing of the master branch and I am trying to make sense of all the data items that are going on. For mean length at age composition data, do other data types depend on having mlacomps? Such that mlacomps are generated even if  ``mlacomp_params`` is null? I have a model with just length and age and I get a message saying that one line of mlacomps were deleted and I am trying to get to why.

**Comment by colemonnahan** [comment]
Hi Kelli, the new `clean_data` function reports that if `verbose==TRUE`. Admittedly I haven't done very thorough testing of it, and at some point we'll want to turn `verbose` off in the base function.

To answer your question, no mla data should be created in your case so this is a bug.

What's probably happening is that I'm representing a blank list element for `mlacomps` than what `r4ss` reads in. I'm passing a NULL but maybe a character string w/ comment would be better? Should be as simple as changing what is hard coded.

**Comment by kellijohnson** [comment]
No actual data is being created, instead ``clean_data`` is reporting the removal of a hash tag, which is generated in ``sample_mlacomp`` by the following line: ``datfile$MeanSize_at_Age_obs <- data.frame("#")``. I think we have talked about this before regarding using a character or NULL and I think we should use NULL since we are using the ``r4ss`` read and write dat functions rather than just ``write_lines``.

**Comment by colemonnahan** [comment]
I think the reason we went with `data.frame("#")` was because that returns 0 rows, whereas `nrow(NULL)` is NULL. But, we can simply use `NROW` and it should behave as expected.

**Comment by colemonnahan** [comment]
I'm working on some test scenarios that are helping me discover bugs. I'll add this to my list and try to get to it soon.

**Comment by kellijohnson** [comment]
@colemonnahan ```sample_acomp```, ran by ```ss3sim_base``` removes potential year / fleet combinations that may be needed later for ```sample_mlacomp```
Should I look into how to fix this or do you know an easy way to fix it? I would think that you could add years and fleets needed for mla comp to sample args then wouldn't ```clean_data``` remove the unnecessary rows later?

**Comment by colemonnahan** [comment]
Hi Kelli, I was under the impression that the mla data should be a strict subset of the agecomp data. I haven't looked carefully at mlacomps yet, but for the others we use the sampled agecomps for the other sampling. So you can't have mla data if you don't have age data. Is this wrong?

**Comment by kellijohnson** [comment]
Probably in reality this is true, but we were wanting to have mlacomp 
data to feed to the external growth estimator. I can also see the 
situation where you might have mean length at age data and not want to 
include the age data because wouldn't that be double counting? @taylori 
do you have any thoughts on this regarding actual assessments.

**Comment by juanlvalero** [comment]
Kelli is right, we are using mean length at age to generate paired
observations of age and length for the external estimator. In real
situations we can have those type of paired observations without having
actual age compositions, in a sense that the ages may not be a
representative sample for composition data, but still useful for estimating
growth. What to do when we have both age comps and length comps is
debatable, some people assign half the weight to each data type (using
lambdas or emphasis factors), some people favor one type of data over the
other.
So in summary, the mla data does not need to be strictly a subset of the
agecomp data. You should be able to have mla data even with no age
composition data. Does this make sense?
Juan

**Comment by colemonnahan** [comment]
Yes this makes sense. So what you're saying is that you need the expected age compositions to be available to sample mla data. This is possible because we can read in the original dat file from the OM and modify it piece by piece, but as a new object. We already do something similar before `change_data` is called (see [here](https://github.com/ss3sim/ss3sim/blob/master/R/ss3sim_base.r#L290)). So the original expected values are always available throughout `ss3sim_base`. You could just use that.

I'm finding it very difficult to generalize the code for all possible cases of data for all groups. I'm starting to think we should just piece the set of data scenarios we need for the growth studies, code them, and move on. The original package only supported age and length comps, so we've already added a ton of functionality.

It's not exactly clear to me what you want to do with mla data, but I'd be happy to sit down with someone and adjust the code accordingly.

**Comment by taylori** [comment]
Juan covered everything that I would have said and more.
I think that regardless of generalized you make the code, the next round of
usage will include new things that require further changes, so I agree with
Cole's sentiment that getting it working adequately for the needs of this
set of papers should be your target for the moment.
-Ian

**Comment by kellijohnson** [comment]
@colemonnahan I just want to make sure I am not missing something, but shouldn't line 114 in ss3sim/R/sample_calcomp.R be
```prob.len.ints <- unlist(sapply(1:length(prob.len), function(i) rep(i, prob.len[i])))```
not 
```prob.len.ints <- unlist(sapply(1:length(prob.len)), function(i) rep(i, prob.len[i]))```

**Comment by kellijohnson** [comment]
Also I am confused on how this works in practice b/c isn't ```prob.len``` supposed to be proportions and not the number of samples per length bin? I would think you would want to:
```r
observedperbin <- prob.len * lencomp[lencomp$Yr==yr & lencomp$FltSvy==fl, ]$NSamp
prob.len.ints <- unlist(sapply(1:length(observedperbin), function(i) rep(i, observedperbin)))
```

**Comment by colemonnahan** [comment]
@kellijohnson Regarding the first question, yes you're right. I thought I fixed that but maybe forgot to push it? Please fix!

Second question: `sample_lcomp` returns integers if the multinomial is used (see [here](https://github.com/ss3sim/ss3sim/blob/master/R/sample_lcomp.R#L128)), otherwise real numbers representing probabilities. We used to convert to proportions but don't anymore, and I can't remember why. Just need to be consistent.

**Comment by kellijohnson** [comment]
1. I will fix
2. ```sample_agecomp``` returns proportions, when using the [multinomial](https://github.com/ss3sim/ss3sim/blob/master/R/sample_agecomp.R#L148) . What should we do? a) leave them to return different things or b) fix ```sample_agecomp```?

**Comment by colemonnahan** [comment]
I think I prefer to return integers from the multinomial. It makes the data file much easier to check/verify and just seems right to me.

At some point I'll switch from the Dirichlet to the Dirichlet-Multinomial (condition the MN prob vector on a Dirichlet draw). That way it always returns integers and there is a more meaningful N. I just haven't worked through the properties of that distribution well enough to convert over. (Anyone wanting to tackle that is welcome by the way).

**Comment by kellijohnson** [comment]
can you think of any repercussions off of the top of your head of returning integers from ```sample_agecomp```. I just ran a quick test and the package still worked but I haven't looked into it too deeply.

**Comment by colemonnahan** [comment]
SS normalizes the data internally, and I can't think of any other issue it would cause. We'll find out!

**Comment by taylori** [comment]
My preference in practice is always to use integers in the composition data
when possible.
In practice, the compositions are often the weighted some of data from
multiple sources and can't be integers, but if you have the option, I agree
with Cole that you should use it.

**Comment by kellijohnson** [comment]
@colemonnahan do you think we have done enough testing with the sampling functions to close this issue?

**Comment by colemonnahan** [comment]
Yes, I have on my plate to describe the updates in the vignette but I'm happy with the state of the code itself.
